src/js/common.ml

```diff
@@ -57,21 +57,35 @@ let js_candid source =
       Js.some (Js.string code)
     )
 
+let js_stable_compatible pre post =
+  js_result (Pipeline.stable_compatible (Js.to_string pre) (Js.to_string post)) (fun _ -> Js.null)
+
 let js_compile_wasm mode source =
+  let source = Js.to_string source in
   let mode =
     match Js.to_string mode with
     | "wasi" -> Flags.WASIMode
     | "ic" -> Flags.ICMode
     | _ -> raise (Invalid_argument "js_compile_with: Unexpected mode")
   in
-  js_result (Pipeline.compile_files mode true [Js.to_string source])
-    (fun (_, m) ->
+  js_result (Pipeline.compile_files mode true [source])
+    (fun (idl_prog, m) ->
+      let open CustomModule in
+      let sig_ = match m.motoko.stable_types with
+        | Some (_, txt) -> Js.some (Js.string txt)
+        | _ -> Js.null in
+      let candid = Idllib.Arrange_idl.string_of_prog idl_prog in
       let (_, wasm) = CustomModuleEncode.encode m in
       let constructor = Js.Unsafe.global##._Uint8Array in
       let code = constructor##from
```

Comment:
I know this old code, but is the code returned as a stream, not just a JS bytearray? Doesn't that make reading each byte really slow?

Reply:
It's returning a bytearray/Uint8Array. The constructor is slow, but it's kind of the only way to produce a bytearray to JS. Other "faster" method breaks the content if it's not utf8 compatible.